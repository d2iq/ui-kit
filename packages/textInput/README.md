# Forms

## TextInput
`TextInput` is a component used to enter text for forms.

### TextInput Appearance & States

**Appearance**
 
 - **Standard**: The default `appearance` prop value is `TextInputAppearance.Standard`.
 - **Error**: The `appearance` prop can be set to `TextInputAppearance.Error`  to render the `TextInput` as invalid. This should be used if the current value has been confirmed invalid.
 - **Success**: The `appearance` prop can be set to `TextInputAppearance.Success` to render the `TextInput` as validated. This should be used if the current value has confirmed valid.

**States**
 
 - **Disabled**: setting the `disabled` prop to `true` will put the `TextInput` into a disabled state. The `TextInput` will be rendered with a grey background and will not be able to take input.
 
### TextInput `inputLabel`

The `inputLabel` prop can be given either a `string` or `ReactNode` to render as a label for the `TextInput` component.

The `showInputLabel` prop can be set to `false` to hide the input's label. This prop defaults to `true`. 

### TextInput `tooltipContent`

The `tooltipContent` prop is a non-required property that can be a `ReactNode`.

If the `tooltipContent` prop is set, an icon tooltip with the given text will be shown next to the label and above the input.

## TextInputWithIcon
`TextInputWithIcon` extends the `TextInput` component but adds two additional props, `iconStart` & `iconEnd`. These props can be provided a `ReactNode` to be rendered either at the start or end of the `TextInput` or both.

## TextInputWithBadges
`TextInputWithBadges` extends the `TextInputWithIcon` component. More info coming soon...

## TextInputWithButtons
`TextInputWithButtons` extends the `TextInputWithIcon` component, does not take an `iconEnd` prop.

A `TextInputWithButtons` component is used when there is an action a user can take that is related to the text input. For example, clicking a button to clear the input's value.

Please do not pass more than 2 buttons into a `TextInputWithButtons`. Instead, consider putting additional buttons outside of the text input.

### Buttons
The `TextInputButton` component is provided to keep text input buttons consistent. To save space, the buttons are always icons. These icon buttons can be filled with any color, and they do not get the same fill color as icons inside the text input that are not buttons.

Because these buttons contain no text, please pass an `aria-label` to the button. For example, a button that clears the text input might get `aria-label="Clear input"`.
