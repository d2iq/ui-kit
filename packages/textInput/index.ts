export { default as TextInput } from "./components/TextInput";
export { default as TextInputWithIcon } from "./components/TextInputWithIcon";
export { default as TextInputWithBadges } from "./components/TextInputWithBadges";
export { default as TextInputWithButtons } from "./components/TextInputWithButtons";
