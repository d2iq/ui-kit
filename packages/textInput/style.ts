import { css } from "@emotion/css";
import { spaceXs } from "../design-tokens/build/js/designTokens";
import { textInputHeight } from "../shared/styles/formStyles";

const badgeGapSize = spaceXs;

export const badgeInputContainer = css`
  height: auto;
  min-height: ${textInputHeight}px;
  padding-bottom: ${parseInt(badgeGapSize, 10) / 2}px;
  padding-top: ${parseInt(badgeGapSize, 10) / 2}px;
`;

export const badgeText = css`
  white-space: nowrap;
`;

export const badgeInputContents = css`
  display: inline-flex;
  flex-direction: column;
  justify-content: center;
  padding-bottom: ${parseInt(badgeGapSize, 10) / 2}px;
  padding-top: ${parseInt(badgeGapSize, 10) / 2}px;
  padding-right: ${badgeGapSize};

  &,
  * {
    max-width: 100%;
  }
`;

export const badgeInput = css`
  flex-grow: 1;
  min-width: 50px;
`;
