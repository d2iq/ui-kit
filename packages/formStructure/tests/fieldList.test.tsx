import React from "react";

import { mount, render } from "enzyme";
import toJSON from "enzyme-to-json";
import {
  FieldList,
  FieldListColumn,
  FieldListColumnSeparator,
  FieldListAddButton
} from "..";
import { TextInput } from "../../textInput";

const mockItems = [
  {
    name: "Brian Vaughn",
    role: "Software Engineer",
    city: "San Jose",
    id: 0
  },
  {
    name: "Jon Doe",
    role: "Product engineer",
    city: "Mountain View",
    id: 1
  },
  {
    name: "Jane Doe",
    role: "UX Designer",
    city: "San Francisco",
    id: 2
  }
];

describe("FieldList", () => {
  it("renders", () => {
    const testRemoveHandler = jest.fn(_rowIndex => () => {
      jest.fn();
    });
    const testFieldUpdateHandler = jest.fn((_rowIndex, _pathToValue) => _e => {
      jest.fn();
    });
    expect(
      toJSON(
        render(
          <FieldList
            data={mockItems}
            onRemoveItem={testRemoveHandler}
            disabledRows={[0]}
            pathToUniqueKey="id"
          >
            <FieldListColumn
              header="Name"
              pathToValue="name"
              onChange={testFieldUpdateHandler}
              key="name"
            >
              {({ defaultProps, onChange, value }) => (
                <TextInput
                  value={value}
                  onChange={onChange}
                  {...defaultProps}
                />
              )}
            </FieldListColumn>
            <FieldListColumnSeparator key="separator">
              :
            </FieldListColumnSeparator>
            <FieldListColumn
              header="Role"
              pathToValue="role"
              onChange={testFieldUpdateHandler}
              key="role"
            >
              {({ defaultProps, onChange, value }) => (
                <TextInput
                  value={value}
                  onChange={onChange}
                  {...defaultProps}
                />
              )}
            </FieldListColumn>
            <FieldListAddButton>Add row</FieldListAddButton>
          </FieldList>
        )
      )
    ).toMatchSnapshot();
  });

  it("calls onAddItem when clicking the add button", () => {
    const newItem = { id: 42 };
    const testAddHandlerInner = jest.fn();
    const testAddHandler = jest.fn(itemToAdd => () => {
      testAddHandlerInner(itemToAdd);
    });
    const testFieldUpdateHandler = jest.fn((_rowIndex, _pathToValue) => _e => {
      jest.fn();
    });
    const fieldListComponent = mount(
      <FieldList
        data={mockItems}
        onAddItem={testAddHandler(newItem)}
        pathToUniqueKey="id"
      >
        <FieldListColumn
          key="name"
          header="Name"
          pathToValue="name"
          onChange={testFieldUpdateHandler}
        >
          {({ defaultProps, onChange, value }) => (
            <TextInput value={value} onChange={onChange} {...defaultProps} />
          )}
        </FieldListColumn>
        <FieldListColumn
          key="role"
          header="Role"
          pathToValue="role"
          onChange={testFieldUpdateHandler}
        >
          {({ defaultProps, onChange, value }) => (
            <TextInput value={value} onChange={onChange} {...defaultProps} />
          )}
        </FieldListColumn>
        <FieldListAddButton>Add Item</FieldListAddButton>
      </FieldList>
    );
    const addButton = fieldListComponent
      .find('button[data-cy="secondaryButton"]')
      .first();

    expect(testAddHandlerInner).not.toHaveBeenCalled();
    addButton.simulate("click");
    expect(testAddHandlerInner).toHaveBeenCalledWith(newItem);
  });

  it("calls onRemoveItem when clicking the remove button", () => {
    const testRemoveHandlerInner = jest.fn();
    const testRemoveHandler = jest.fn(rowIndex => () => {
      testRemoveHandlerInner(rowIndex);
    });
    const testFieldUpdateHandler = jest.fn((_rowIndex, _pathToValue) => _e => {
      jest.fn();
    });
    const fieldListComponent = mount(
      <FieldList
        data={mockItems}
        onRemoveItem={testRemoveHandler}
        pathToUniqueKey="id"
      >
        <FieldListColumn
          key="name"
          header="Name"
          pathToValue="name"
          onChange={testFieldUpdateHandler}
        >
          {({ defaultProps, onChange, value }) => (
            <TextInput value={value} onChange={onChange} {...defaultProps} />
          )}
        </FieldListColumn>
        <FieldListColumn
          key="role"
          header="Role"
          pathToValue="role"
          onChange={testFieldUpdateHandler}
        >
          {({ defaultProps, onChange, value }) => (
            <TextInput value={value} onChange={onChange} {...defaultProps} />
          )}
        </FieldListColumn>
      </FieldList>
    );
    const removeButton = fieldListComponent
      .find('button[data-cy="fieldList-removeButton"]')
      .first();

    expect(testRemoveHandlerInner).not.toHaveBeenCalled();
    removeButton.simulate("click");
    expect(testRemoveHandlerInner).toHaveBeenCalledWith(0);
  });

  it("disables the inputs and buttons in the rows specified by the disabledRows prop", () => {
    const testRemoveHandlerInner = jest.fn();
    const testRemoveHandler = jest.fn(rowIndex => () => {
      testRemoveHandlerInner(rowIndex);
    });
    const testFieldUpdateHandler = jest.fn((_rowIndex, _pathToValue) => _e => {
      jest.fn();
    });
    const fieldListComponent = mount(
      <FieldList
        data={mockItems}
        onRemoveItem={testRemoveHandler}
        disabledRows={[0]}
        pathToUniqueKey="id"
      >
        <FieldListColumn
          key="name"
          header="Name"
          pathToValue="name"
          onChange={testFieldUpdateHandler}
        >
          {({ disabled, defaultProps, onChange, value }) => (
            <TextInput
              disabled={disabled}
              value={value}
              onChange={onChange}
              {...defaultProps}
            />
          )}
        </FieldListColumn>
        <FieldListColumn
          key="role"
          header="Role"
          pathToValue="role"
          onChange={testFieldUpdateHandler}
        >
          {({ disabled, defaultProps, onChange, value }) => (
            <TextInput
              disabled={disabled}
              value={value}
              onChange={onChange}
              {...defaultProps}
            />
          )}
        </FieldListColumn>
      </FieldList>
    );
    const textInputs = fieldListComponent.find(
      'input[data-cy*="textInput-input"]'
    );
    const removeButton = fieldListComponent
      .find('button[data-cy="fieldList-removeButton"]')
      .first();

    textInputs.forEach((input, i) => {
      if (i < 2) {
        expect(input.prop("disabled")).toBe(true);
      }
      if (i > 1) {
        expect(input.prop("disabled")).toBe(false);
      }
    });
    expect(removeButton.prop("disabled")).toBe(true);
  });

  it("calls FieldListColumn's `onChange` render prop", () => {
    const pathToValue = "name";
    const testFieldUpdateHandlerInner = jest.fn();
    const testRemoveHandler = jest.fn(_rowIndex => () => {
      jest.fn();
    });
    const testFieldUpdateHandler = jest.fn((rowIndex, pathToValue) => _e => {
      testFieldUpdateHandlerInner(rowIndex, pathToValue);
    });
    const fieldListComponent = mount(
      <FieldList
        data={mockItems}
        onRemoveItem={testRemoveHandler}
        pathToUniqueKey="id"
      >
        <FieldListColumn
          key={pathToValue}
          header="Name"
          pathToValue={pathToValue}
          onChange={testFieldUpdateHandler}
        >
          {({ defaultProps, onChange, value }) => (
            <TextInput value={value} onChange={onChange} {...defaultProps} />
          )}
        </FieldListColumn>
      </FieldList>
    );

    const textInput = fieldListComponent
      .find('input[data-cy*="textInput-input"]')
      .first();

    expect(testFieldUpdateHandlerInner).not.toHaveBeenCalled();
    textInput.simulate("change", { value: "D2iQ" });
    expect(testFieldUpdateHandlerInner).toHaveBeenCalledWith(0, pathToValue);
  });
});
