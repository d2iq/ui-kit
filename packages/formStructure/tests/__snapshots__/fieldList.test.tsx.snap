// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldList renders 1`] = `
<div>
  <div
    class="css-1olmsyh"
    data-cy="spacingBox"
  >
    <div
      class="rmTextMargins css-49hn1i"
      data-cy="fieldList-columnHeader"
    >
      Name
    </div>
    <div
      class="rmTextMargins css-4tue1l"
      data-cy="fieldList-columnHeader"
    >
      :
    </div>
    <div
      class="rmTextMargins css-49hn1i"
      data-cy="fieldList-columnHeader"
    >
      Role
    </div>
  </div>
  <div
    class="css-jq21gt"
  >
    <div
      class="css-1wztphk"
    >
      <div
        class="css-y6q8rd"
        data-cy="fieldList-cell"
      >
        <div
          data-cy="textInput textInput.standard"
        >
          <label
            class="css-1lrhbps"
            data-cy="textInput-label"
            for="name-0"
          >
            Name 0
          </label>
          <div>
            <div
              class="css-865tln"
            >
              <input
                aria-describedby=""
                aria-invalid="false"
                class="css-1w8it3t"
                data-cy="textInput-input"
                id="name-0"
                type="text"
                value="Brian Vaughn"
              />
            </div>
            <div
              data-cy="textInput-hintContent"
            />
          </div>
        </div>
      </div>
      <div
        class="css-y6q8rd"
        data-cy="fieldList-cell"
      >
        :
      </div>
      <div
        class="css-y6q8rd"
        data-cy="fieldList-cell"
      >
        <div
          data-cy="textInput textInput.standard"
        >
          <label
            class="css-1lrhbps"
            data-cy="textInput-label"
            for="role-0"
          >
            Role 0
          </label>
          <div>
            <div
              class="css-865tln"
            >
              <input
                aria-describedby=""
                aria-invalid="false"
                class="css-1w8it3t"
                data-cy="textInput-input"
                id="role-0"
                type="text"
                value="Software Engineer"
              />
            </div>
            <div
              data-cy="textInput-hintContent"
            />
          </div>
        </div>
      </div>
      <div>
        <button
          class="css-gvfau5"
          data-cy="fieldList-removeButton"
          disabled=""
          type="button"
        >
          <div
            tabindex="-1"
          >
            <svg
              aria-label="system-close icon"
              class="css-xp6z5l"
              data-cy="icon"
              height="16"
              preserveAspectRatio="xMinYMin meet"
              role="img"
              viewBox="0 0 16 16"
              width="16"
            >
              <use
                href="#system-close"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
    <div
      class="css-1wztphk"
    >
      <div
        class="css-y6q8rd"
        data-cy="fieldList-cell"
      >
        <div
          data-cy="textInput textInput.standard"
        >
          <label
            class="css-1lrhbps"
            data-cy="textInput-label"
            for="name-1"
          >
            Name 1
          </label>
          <div>
            <div
              class="css-865tln"
            >
              <input
                aria-describedby=""
                aria-invalid="false"
                class="css-1w8it3t"
                data-cy="textInput-input"
                id="name-1"
                type="text"
                value="Jon Doe"
              />
            </div>
            <div
              data-cy="textInput-hintContent"
            />
          </div>
        </div>
      </div>
      <div
        class="css-y6q8rd"
        data-cy="fieldList-cell"
      >
        :
      </div>
      <div
        class="css-y6q8rd"
        data-cy="fieldList-cell"
      >
        <div
          data-cy="textInput textInput.standard"
        >
          <label
            class="css-1lrhbps"
            data-cy="textInput-label"
            for="role-1"
          >
            Role 1
          </label>
          <div>
            <div
              class="css-865tln"
            >
              <input
                aria-describedby=""
                aria-invalid="false"
                class="css-1w8it3t"
                data-cy="textInput-input"
                id="role-1"
                type="text"
                value="Product engineer"
              />
            </div>
            <div
              data-cy="textInput-hintContent"
            />
          </div>
        </div>
      </div>
      <div>
        <button
          class="css-1smnoyq"
          data-cy="fieldList-removeButton"
          type="button"
        >
          <div
            tabindex="-1"
          >
            <svg
              aria-label="system-close icon"
              class="css-k06270"
              data-cy="icon"
              height="16"
              preserveAspectRatio="xMinYMin meet"
              role="img"
              viewBox="0 0 16 16"
              width="16"
            >
              <use
                href="#system-close"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
    <div
      class="css-1wztphk"
    >
      <div
        class="css-y6q8rd"
        data-cy="fieldList-cell"
      >
        <div
          data-cy="textInput textInput.standard"
        >
          <label
            class="css-1lrhbps"
            data-cy="textInput-label"
            for="name-2"
          >
            Name 2
          </label>
          <div>
            <div
              class="css-865tln"
            >
              <input
                aria-describedby=""
                aria-invalid="false"
                class="css-1w8it3t"
                data-cy="textInput-input"
                id="name-2"
                type="text"
                value="Jane Doe"
              />
            </div>
            <div
              data-cy="textInput-hintContent"
            />
          </div>
        </div>
      </div>
      <div
        class="css-y6q8rd"
        data-cy="fieldList-cell"
      >
        :
      </div>
      <div
        class="css-y6q8rd"
        data-cy="fieldList-cell"
      >
        <div
          data-cy="textInput textInput.standard"
        >
          <label
            class="css-1lrhbps"
            data-cy="textInput-label"
            for="role-2"
          >
            Role 2
          </label>
          <div>
            <div
              class="css-865tln"
            >
              <input
                aria-describedby=""
                aria-invalid="false"
                class="css-1w8it3t"
                data-cy="textInput-input"
                id="role-2"
                type="text"
                value="UX Designer"
              />
            </div>
            <div
              data-cy="textInput-hintContent"
            />
          </div>
        </div>
      </div>
      <div>
        <button
          class="css-1smnoyq"
          data-cy="fieldList-removeButton"
          type="button"
        >
          <div
            tabindex="-1"
          >
            <svg
              aria-label="system-close icon"
              class="css-k06270"
              data-cy="icon"
              height="16"
              preserveAspectRatio="xMinYMin meet"
              role="img"
              viewBox="0 0 16 16"
              width="16"
            >
              <use
                href="#system-close"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
    <button
      class="css-1qbr1ah"
      data-cy="secondaryButton"
      tabindex="0"
      type="button"
    >
      <span
        class="css-1hesotu"
      >
        <span
          class="css-16x89v1"
        >
          <svg
            aria-label="system-plus icon"
            class="css-1naj4ae"
            data-cy="icon"
            height="16"
            preserveAspectRatio="xMinYMin meet"
            role="img"
            viewBox="0 0 16 16"
            width="16"
          >
            <use
              href="#system-plus"
            />
          </svg>
        </span>
        <span
          class="css-mzjeix"
        >
          Add row
        </span>
      </span>
    </button>
  </div>
</div>
`;
